src/object_store/src/object/opendal_engine/opendal_object_store.rs

```diff
@@ -0,0 +1,386 @@
+        fail_point!("opendal_streaming_read_err", |_| Err(
+            ObjectError::internal("opendal streaming read error")
+        ));
+
+        let mut reader: ObjectReader = self.op.object(path).reader().await?;
+        let bytes = match start_pos {
+            Some(start_position) => {
+                let mut buf = Vec::new();
+
+                reader.seek(SeekFrom::Start(start_position as u64)).await?;
+                reader.read_to_end(&mut buf).await?;
+                Bytes::from(buf)
+            }
+            None => {
+                let mut buf = Vec::new();
+                reader.read_to_end(&mut buf).await?;
+                Bytes::from(buf)
+            }
+        };
+
+        Ok(Box::new(Cursor::new(bytes)))
+    }
+
+    async fn metadata(&self, path: &str) -> ObjectResult<ObjectMetadata> {
+        let opendal_metadata = self.op.object(path).metadata().await?;
+        let key = path.to_string();
+        let last_modified = match opendal_metadata.last_modified() {
+            Some(t) => t.unix_timestamp() as f64,
+            None => 0_f64,
+        };
+
+        let total_size = opendal_metadata.content_length() as usize;
+        let metadata = ObjectMetadata {
+            key,
+            last_modified,
+            total_size,
+        };
+        Ok(metadata)
+    }
+
+    async fn delete(&self, path: &str) -> ObjectResult<()> {
+        self.op.object(path).delete().await?;
+        Ok(())
+    }
+
+    /// Deletes the objects with the given paths permanently from the storage. If an object
+    /// specified in the request is not found, it will be considered as successfully deleted.
+    async fn delete_objects(&self, paths: &[String]) -> ObjectResult<()> {
+        for path in paths {
+            self.op.object(path).delete().await?;
+        }
+        Ok(())
+    }
+
+    async fn list(&self, prefix: &str) -> ObjectResult<Vec<ObjectMetadata>> {
+        let mut object_lister = self.op.object(prefix).list().await?;
+        let mut metadata_list = vec![];
+        while let Some(obj) = object_lister.next().await {
+            let object = obj?;
+            let key = prefix.to_string();
+            let last_modified = match object.metadata().await?.last_modified() {
+                Some(t) => t.unix_timestamp() as f64,
+                None => 0_f64,
+            };
+
+            let total_size = object.metadata().await?.content_length() as usize;
+            let metadata = ObjectMetadata {
+                key,
+                last_modified,
+                total_size,
+            };
+            metadata_list.push(metadata);
+        }
+        Ok(metadata_list)
+    }
+
+    fn store_media_type(&self) -> &'static str {
+        match self.engine_type {
+            EngineType::Memory => "Memory",
+            EngineType::Hdfs => "Hdfs",
+        }
+    }
+}
+
+/// Store multiple parts in a map, and concatenate them on finish.
+pub struct HdfsStreamingUploader {
```

Comment:
This has nothing to do with HDFS so it is better to name it OpenDALStreamingUploader instead.